graph/lowest_common_ancestor.cpp

```diff
@@ -0,0 +1,247 @@
+/**
+ *
+ * \file
+ *
+ * \brief Data structure for finding the lowest common ancestor
+ * of two vertices in a rooted tree using binary lifting.
+ *
+ * \details
+ * Algorithm: https://cp-algorithms.com/graph/lca_binary_lifting.html
+ *
+ * Complexity:
+ *   - Precomputation: O(N log N) where N is the number of vertices in the tree
+ *   - Query: O(log N)
+ *   - Space: O(N log N)
+ *
+ * Example:
+ * <br/>  Tree:
+ * <pre>
+ *             _  3  _
+ *          /     |     \
+ *        1       6       4
+ *      / |     /   \       \
+ *    7   5   2       8       0
+ *            |
+ *            9
+ * </pre>
+ *
+ * <br/>  lowest_common_ancestor(7, 4) = 3
+ * <br/>  lowest_common_ancestor(9, 6) = 6
+ * <br/>  lowest_common_ancestor(0, 0) = 0
+ * <br/>  lowest_common_ancestor(8, 2) = 6
+ *
+ *   The query is symmetrical, therefore
+ *     lowest_common_ancestor(x, y) = lowest_common_ancestor(y, x)
+ */
+
+#include <iostream>
+#include <utility>
+#include <vector>
+#include <queue>
+#include <cassert>
+
+/**
+ * Class for representing a graph as an adjacency list.
+ * Its vertices are indexed 0, 1, ..., N - 1.
+ */
+class Graph {
+  public:
+    /**
+     * \brief Populate the adjacency list for each vertex in the graph.
+     * Assumes that evey edge is a pair of valid vertex indices.
+     *
+     * @param N number of vertices in the graph
+     * @param undirected_edges list of graph's undirected edges
+     */
+    Graph(unsigned long N, const std::vector< std::pair<int, int> > &undirected_edges) {
+        neighbors.resize(N);
+        for (auto &edge : undirected_edges) {
+            neighbors[edge.first].push_back(edge.second);
+            neighbors[edge.second].push_back(edge.first);
+        }
+    }
+
+    /**
+     * @return the number of vertices in the graph.
```

Comment:
```suggestion
     * Function to get the number of vertices in the graph
     * @return the number of vertices in the graph.
```
I know this is redundant, but `doxygen` sometimes fails to interpret the functions if the descriptions do not have a `@param` or `@returns`